### PR TITLE
Make CRUD API + pagination treatment consistent across resources (2)

### PR DESCRIPTION
**Pull request by SublimeDamien**
https://github.com/cloudflare/cloudflare-go/pull/138

### CODE REVIEW

**Comment by SublimeDamien** [comment]
@elithrar Can you trigger the CI please ?

**Comment by elithrar** [comment]
Travis failed as it seems to have detected 'abuse' (it doesn't say what triggered that determination).

However, there's a lot of changes in here, and the commit messages don't line up. Can you rebase these commits to provide clearer commit messages?

**Comment by SublimeDamien** [comment]
Ok I'll do that next week

**Comment by elithrar** [comment]
@SublimeDamien Let me know when you're ready for me to re-review.

**Comment by SublimeDamien** [comment]
Hi @elithrar I was little busy.
I split the commit in multiple commits

**Comment by SublimeDamien** [comment]
@elithrar I have an error on the following command `diff -u <(echo -n) <(gofmt -d .)` but I don't understand the goal of this command and why there is a failure

**Comment by mwarkentin** [comment]
@SublimeDamien Not 100%, but I think that's a check to make sure that code has been run through `gofmt` before check in? Might just need to run that first?

**Comment by elithrar** [comment]
Correct, it’s gofmt.

**Comment by elithrar** [comment]
Thanks @SublimeDamien - going to review later today!

**Comment by SublimeDamien** [comment]
Ok, you're welcome.
I'm waiting your reviews

**Comment by elithrar** [comment]
For the other reviewers, this PR *should* - 

* Allow a package user to paginate appropriately, depending on whether they find their results vs. network calls (and thus, time). Users with thousands of records would not want auto-pagination.
* Make method signatures as consistent as possible (pointer vs. value return, pointer vs. value args, etc) - this may require some bikeshedding (proposal: value for arg to avoid mutation, pointer for response to minimize copying large responses around)
* Return responses for all methods to allow the caller to retrieve IDs as necessary. The PageRules methods were the largest culprit here.

It will break the public API, and thus we'll bump the major version after merging.

File: cloudflare.go [hunk 1]
```diff
@@ -67,16 +67,17 @@ func (api *API) SetAuthType(authType int) {
 
 // ZoneIDByName retrieves a zone's ID from the name.
 func (api *API) ZoneIDByName(zoneName string) (string, error) {
-	res, err := api.ListZones(zoneName)
+	//TODO: paginate
```
Comment (elithrar) [inline cloudflare.go]: Remove this comment, as not necessary. The FilterZones call should search for the exact name via the `name=${name}` param.

File: cmd/flarectl/dns.go [hunk 2]
```diff
@@ -90,18 +90,18 @@ func dnsCreateOrUpdate(c *cli.Context) {
 	rr := cloudflare.DNSRecord{
 		Name: name + "." + zone,
 	}
-	records, err := api.DNSRecords(zoneID, rr)
+	records, err := api.FilterDNSRecords(zoneID, 0, rr)
```
Comment (elithrar) [inline cmd/flarectl/dns.go]: Why do we set page=0 here? The underlying call seems to coerce 0 to 1, but for clarity we should call page=1, which is what the HTTP API accepts.

File: cmd/flarectl/firewall.go [hunk 3]
```diff
@@ -320,12 +320,12 @@ func getScope(c *cli.Context) (string, string, error) {
 	var organization, organizationID string
 	if c.String("organization") != "" {
 		organization = c.String("organization")
-		organizations, _, err := api.ListOrganizations()
+		organizations, err := api.ListOrganizations(0)
```
Comment (elithrar) [inline cmd/flarectl/firewall.go]: Same here. Why not `ListOrganizations(1)`

File: cmd/flarectl/misc.go [hunk 4]
```diff
@@ -61,14 +61,14 @@ func pageRules(c *cli.Context) {
 		return
 	}
 
-	rules, err := api.ListPageRules(zoneID)
+	rules, err := api.ListPageRules(zoneID, 0)
```
Comment (elithrar) [inline cmd/flarectl/misc.go]: `page=1`

File: cmd/flarectl/zone.go [hunk 5]
```diff
@@ -64,13 +64,13 @@ func zoneList(c *cli.Context) {
 		fmt.Println(err)
 		return
 	}
-	zones, err := api.ListZones()
+	zones, err := api.ListZones(0)
```
Comment (elithrar) [inline cmd/flarectl/zone.go]: `page=1`

**Comment by SublimeDamien** [comment]
@elithrar There is other things to change ?

**Comment by elithrar** [comment]
No sorry! Busy week here. Will make an effort to review and hopefully
merge soon.

(It’s a large PR with multiple breaking changes and thus requires more
eyeballs)

File: zone_example_test.go [hunk 6]
```diff
@@ -31,7 +31,7 @@ func ExampleAPI_ListZones_filter() {
 	}
 
 	// Fetch a slice of zones example.org and example.net.
-	zones, err := api.ListZones(1, "example.org", "example.net")
```
Comment (cbranch) [inline zone_example_test.go]: This commit should be removed:

* you fix the call in a later commit
* your change no longer matches the comment

**Comment by SublimeDamien** [comment]
@cbranch I tried to clean up, say me if it's OK

**Comment by SublimeDamien** [comment]
@cbranch @elithrar Hi guys, what is missing for merge this PR ?

**Comment by benjvi** [comment]
I've been reading through this and I'm not sure what the desired benefit is for returning response objects from Create methods. It seems simpler to pass just the object and error back, if that is possible. Is it the case that some resources don't have an ID and so passing the resource object back is insufficient to do lookups? (all the resource types I looked at so far do have IDs). I might be missing something else? @elithrar ?

**Comment by elithrar** [comment]
@benjvi - I'm OK with your suggested approach here. The approach in #169 LGTM.

**Comment by SublimeDamien** [comment]
@elithrar @benjvi If I understand what you want, for an example

actually there is:
```go
func (api *API) CreateCustomHostname(zoneID string, ch CustomHostname) (*CustomHostnameResponse, error) {
	uri := "/zones/" + zoneID + "/custom_hostnames"
	res, err := api.makeRequest("POST", uri, ch)
	if err != nil {
		return nil, errors.Wrap(err, errMakeRequestError)
	}

	var response CustomHostnameResponse
	err = json.Unmarshal(res, &response)
	if err != nil {
		return nil, errors.Wrap(err, errUnmarshalError)
	}

	return &response, nil
}
```

to

```go
func (api *API) CreateCustomHostname(zoneID string, ch CustomHostname) (*CustomHostname, error) {
	uri := "/zones/" + zoneID + "/custom_hostnames"
	res, err := api.makeRequest("POST", uri, ch)
	if err != nil {
		return nil, errors.Wrap(err, errMakeRequestError)
	}

	var response CustomHostnameResponse
	err = json.Unmarshal(res, &response)
	if err != nil {
		return nil, errors.Wrap(err, errUnmarshalError)
	}

	return &response.Result, nil
}
```

**Comment by SublimeDamien** [comment]
@elithrar @benjvi If I understand what you want, for an example

actually there is:
```go
func (api *API) CreateCustomHostname(zoneID string, ch CustomHostname) (*CustomHostnameResponse, error) {
	uri := "/zones/" + zoneID + "/custom_hostnames"
	res, err := api.makeRequest("POST", uri, ch)
	if err != nil {
		return nil, errors.Wrap(err, errMakeRequestError)
	}

	var response CustomHostnameResponse
	err = json.Unmarshal(res, &response)
	if err != nil {
		return nil, errors.Wrap(err, errUnmarshalError)
	}

	return &response, nil
}
```

to

```go
func (api *API) CreateCustomHostname(zoneID string, ch CustomHostname) (*CustomHostname, error) {
	uri := "/zones/" + zoneID + "/custom_hostnames"
	res, err := api.makeRequest("POST", uri, ch)
	if err != nil {
		return nil, errors.Wrap(err, errMakeRequestError)
	}

	var response CustomHostnameResponse
	err = json.Unmarshal(res, &response)
	if err != nil {
		return nil, errors.Wrap(err, errUnmarshalError)
	}

	return &response.Result, nil
}
```

**Comment by benjvi** [comment]
@SublimeDamien yes thats how I see it, passing just the pointer to the object and an error

**Pull request closed without merging**